### ISSUE TITLE
Please implement the neighbors argument

### ISSUE THREAD

**Issue by wkirgsn** [ArmanMaesumi/torchrbf#3]
I would really appreciate the implementation of the neighbors argument in the RBFInterpolator init function

**Comment by Nemo-Malhomme** [comment]
Hello, I would like to also manifest interest in this functionality being implemented.

The current behavior of RBFInterpolation makes it very memory demanding for applications on meshes with a large number of points. Typically, in sufficiently large or dense meshes, there is not need for each interpolated point to use information from every single datapoint. The "neighbor" argument functionality should allow for lower memory and computation cost at initialization, and increase potential use cases.

**Comment by ArmanMaesumi** [comment]
Implementing neighbors functionality poses some challenges

For an efficient (non-quadratic) implementation, we would need to construct a KDTree and reuse it at every RBFInterpolator query. After a cursory search, I only found one satisfactory PyTorch KDTree implementation, which likely complicates building our package. Especially because the spatial dimensionality of the interpolator needs to be know at compile time..
[https://github.com/thomgrand/torch_kdtree](https://github.com/thomgrand/torch_kdtree)

Perhaps it is most straightforward to fallback to CPU and use Scipy's KDTree here. This would also reduce bandwidth for very large point queries.

Does that sound reasonable for your applications?

**Comment by Nemo-Malhomme** [comment]
Thank your for your reply.

In the context of my application, it is required that an interpolator query both preserves, and can be added to, Pytorch's computational graph for the computation of gradients. From my experience in other contexts, I believe that the Scipy implementation generally isn't compatible with Pytorch's gradient tracking, so unless it is in the case of KDTrees, I would prefer a fully Pytorch implementation. Needing to know interpolator spatial dimensionality at compile time is unpractical but only a minor downside.

However, this is all only for my own use case, and it may not reflect what would be most useful to a majority of users. I fully understand that reducing bandwidth for large interpolation queries is desirable.

**Comment by ArmanMaesumi** [comment]
I see. Yes autodiff is another concern, which is also a headache even if we used the PyTorch KDTree that I linked -- I suspect it doesn't give you autodiff functionality at all. 

It's worth noting that your pipeline may still be differentiable even when the KDTree isn't. If your optimizer is not updating data coordinates (the ones used to construct KDTree) or query coordinates, then the compute graph doesn't need gradients from the KDTree.

That being said, another option would be to use PyTorch3D's KNN routine, which as far as I know supports autograd:
[https://pytorch3d.readthedocs.io/en/latest/modules/ops.html#pytorch3d.ops.knn_points](https://pytorch3d.readthedocs.io/en/latest/modules/ops.html#pytorch3d.ops.knn_points)

 I also want to point out that differentiating through these sorts of operations can be tricky and you may not get meaningful gradients. There might exist softer forms of KNN that are made specifically for autograd, but I'd need to look around.

**Comment by Nemo-Malhomme** [comment]
Thank you for the advice.
I thought that a radial basis function-based interpolator would naturally come with differentiability of the interpolated variables with respect to the spatial coordinates, but it is true that preserving this property through a selection of nearest neighbors comes with significant complications on a theoretical level.

I will look into what you have proposed and search for alternatives on my own. Thank you for your time.

**Comment by Nemo-Malhomme** [comment]
After reflecting on this problem some more ;

Do you believe it would be feasible to implement the feature in such a way that, during interpolation query, the neighbor selection procedure using KDTrees is ignored by autodiff, but the computation of the interpolated values knowing which neighbors to use are not ? I assume that this is possible, for example by detaching all relevant tensors from the computational graph before the KDTree selection, and then reattaching them by re-adding the graph as it was, before applying the resulting interpolant.

This way would ensure differentiability with respect to the spatial coordinates, while avoiding the problem of meaningless gradients. This would also allow for the use of Scipy's KDTree implementation without loss of applicability.

I understand this would cause gradients to be discontinuous at some locations, which are the same at which the interpolated variable can be discontinuous. This would also make it so the gradient is more of an approximation that is only locally true, rather than an absolute description of the influence of slightly changing the spatial coordinates as autodiff aims to be. However, this would be largely sufficient, and quite helpful, in the context of my application.

**Comment by ArmanMaesumi** [comment]
This depends on if the compute graph needs gradients from the KDTree or not. If you are optimizing the query points for example, then doing this would crash the script (I believe) since the gradients are missing.

In the case where gradients for the KDTree are not needed, then you don't even need to detach the tensors. It should "just work", even if you use a external package like SciPy to compute the KDTree.

**Comment by Nemo-Malhomme** [comment]
The compute graph need not include the gradients for The KDTree itself, but I do need the interpolation process to be included in some way, at least specifically the application of the RBF. I believe it is fine if the KDTree neighbor selection could be "skipped" in the compute graph. From what I understand, gradients of this process would be flat, discontinuous, and not very meaningful or usable.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
However, please note that I can only speak to what would be useful to me, and I do not know what would be better for the highest number of people.

I have never used Scipy's KDTrees specifically, but in my experience, when Scipy needs to do something that cannot be done on Pytorch tensors, it converts them to numpy arrays. This results in the compute graph being completely lost. If it does work as you say, however, then I would indeed be glad to have exactly that as part of the neighbor argument feature.

**Comment by wkirgsn** [comment]
I assume the KDtree nearest neighbors selection can be skipped in the computation graph the same way mini batch training does not compute gradients on how the mini batch is selected. 
A scipy KDtree might be applicable. 

I personally would need only the GPU acceleration part

**Comment by ArmanMaesumi** [comment]
> From what I understand, gradients of this process would be flat, discontinuous, and not very meaningful or usable.

Yes this sounds right to me. I think it's reasonable to support this feature with the assumption that using neighbors has limited differentiability, so all of this seems fine with me. 

I have a local version working now using scipy's KDTree. I'll need a bit of time to polish and test changes though.

**Comment by ArmanMaesumi** [comment]
I have just pushed torchrbf version 1.0.0, which includes neighbors based interpolation, for details see: [v.1.0.0](https://github.com/ArmanMaesumi/torchrbf/releases/tag/1.0.0). Please feel free to reopen this issue if the neighbors functionality has issues. 

I've tested against SciPy's interpolator and the results seem to match up perfectly.
